tools/swift-reflection-dump/swift-reflection-dump.cpp

```diff
@@ -32,200 +33,227 @@
+    for (const ObjectFile *O : ObjectFiles)
+      Images.emplace_back(O);
   }
 
+  const std::vector<Image> &getImages() const { return Images; }
+
   bool queryDataLayout(DataLayoutQueryType type, void *inBuffer,
                        void *outBuffer) override {
     switch (type) {
-      case DLQ_GetPointerSize: {
-        auto result = static_cast<uint8_t *>(outBuffer);
-        *result = sizeof(void *);
-        return true;
-      }
-      case DLQ_GetSizeSize: {
-        auto result = static_cast<uint8_t *>(outBuffer);
-        *result = sizeof(size_t);
-        return true;
-      }
+    case DLQ_GetPointerSize: {
+      auto result = static_cast<uint8_t *>(outBuffer);
+      *result = sizeof(void *);
+      return true;
+    }
+    case DLQ_GetSizeSize: {
+      auto result = static_cast<uint8_t *>(outBuffer);
+      *result = sizeof(size_t);
+      return true;
+    }
     }
 
     return false;
   }
 
   RemoteAddress getSymbolAddress(const std::string &name) override {
-    for (auto &object : ObjectFiles) {
-      for (auto &symbol : object->symbols()) {
-        if (unwrap(symbol.getName()).equals(name)) {
-          // TODO: Account for offset in ELF binaries
-          return RemoteAddress(unwrap(symbol.getAddress()));
-        }
-      }
-    }
     return RemoteAddress(nullptr);
   }
-  
-  bool isAddressValid(RemoteAddress addr, uint64_t size) const {
-    // TODO: Account for offset in ELF binaries
-
-    auto src = addr.getAddressData();
-    
-    // Check that the source is in bounds of one of the object files.
-    for (auto &object : ObjectFiles) {
-      if ((uint64_t)object->getData().bytes_begin() <= src
-          && src + size <= (uint64_t)object->getData().bytes_end()) {
-        return true;
-      }
-    }
-    return false;
-  }
-  
-  ReadBytesResult readBytes(RemoteAddress address, uint64_t size) override {
-    if (!isAddressValid(address, size))
-      return ReadBytesResult(nullptr, [](const void *){});
 
-    // TODO: Account for offset in ELF binaries
-    return ReadBytesResult((const void *)address.getAddressData(), [](const void *) {});
+  ReadBytesResult readBytes(RemoteAddress Addr, uint64_t Size) override {
+    auto I = std::find_if(Images.begin(), Images.end(), [=](const Image &I) {
+      return I.isAddressValid(Addr, Size);
+    });
+    return I == Images.end() ? ReadBytesResult(nullptr, [](const void *) {})
+                             : I->readBytes(Addr, Size);
   }
-  
-  bool readString(RemoteAddress address, std::string &dest) override {
-    if (!isAddressValid(address, 1))
+
+  bool readString(RemoteAddress Addr, std::string &Dest) override {
+    ReadBytesResult R = readBytes(Addr, 1);
+    if (!R)
       return false;
-    // TODO: Account for running off the edge of an object, offset in ELF
-    // binaries
-    auto cString = StringRef((const char*)address.getAddressData());
-    dest.append(cString.begin(), cString.end());
+    StringRef Str((const char *)R.get());
```

Comment:
Hmm, should we pass the length?  Cause otherwise it will get the wrong length (truncate to first NUL byte)

Reply:
this methods overrides the corresponding method from the base class, thus we can't change its signature. I think it is for C-style strings

Reply:
It is trying to read a C string, but as the old TODO notes, though, there is a risk of running off the end if there is a missing terminator. It might be good to account for the size of the mapping when scanning ahead to a null terminator. Maybe we can do that as a follow-up patch.

Reply:
Yeah, that sounds pretty good to me.

Reply:
It looks like the reflection tests are crashing in PR testing in strlen; could it be related to this change?

Reply:
No, its related to trying to do a string compare on the section header name which is not mapped in (as is the expected behaviour with ELF).  Rather than trying to find the section by name, I think that we should mark the sections with flags to indicate the contents (e.g. 0x800000051 for file metadata).  The flags range 0x80000000-0xffffffff is reserved for the application, so I think that this is squarely in the intended usage.